### PR TITLE
fix(lib-collision): tolerate `public_name` collisions in different contexts

### PR DESCRIPTION
**Pull request by anmonteiro**
- fixes the test added in https://github.com/ocaml/dune/pull/10471
- we use the same strategy for public libraries that we use elsewhere in `scope.ml`: we store and `enabled` `Toggle.t Memo.t` that we evaluate when resolving a public library by its name.

### CODE REVIEW

**Review by jchavarri — APPROVED** [review]
Thanks!

File: src/dune_rules/scope.ml [hunk 1]
```diff
@@ -473,15 +460,22 @@ module DB = struct
              | Some lib ->
                let name = Package.name pkg in
                (name, Lib_entry.Library (Lib.Local.of_lib_exn lib)) :: acc)
-          | Library.T { visibility = Public pub; _ } ->
-            let+ lib = Lib.DB.find public_libs (Public_lib.name pub) in
+          | Library.T { visibility = Public pub; enabled_if; _ } ->
```
Comment (rgrinberg) [inline src/dune_rules/scope.ml:463]: Why don't private libraries have an enabled_if check?

File: src/dune_rules/scope.ml [hunk 2]
```diff
@@ -473,15 +460,22 @@ module DB = struct
...
+            let* lib = Lib.DB.find public_libs (Public_lib.name pub) in
```
Comment (rgrinberg) [inline src/dune_rules/scope.ml:464]: You should use the id rather than the name for this check.

File: src/dune_rules/scope.ml (hunk 1 above)
Comment (anmonteiro) [reply src/dune_rules/scope.ml:463]: that's a good point. I'll make these changes shortly.

File: src/dune_rules/scope.ml (hunk 2 above)
Comment (anmonteiro) [reply src/dune_rules/scope.ml:464]: I tried `find_by_lib_id` but I got a crash about adding a duplicate name in https://github.com/ocaml/dune/blob/main/src/dune_rules/install_rules.ml#L697

Comment (anmonteiro) [reply src/dune_rules/scope.ml:464]: I'm not sure what that means yet.

Comment (anmonteiro) [reply src/dune_rules/scope.ml:464]: Ah, it's likely because it's finding the right lib now.

**Pull request merged**